README.md
# hyper-simulator

[![Build Status](https://travis-ci.org/decentstack/hyper-simulator.svg?branch=master)](https://travis-ci.org/decentstack/hyper-simulator)
[![JavaScript Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://standardjs.com)

> Swarm Integration Testing - because unit-tests for P2P sucks.

This project grew out of the frustration of always finding new and exotic bugs in production that were a complete pain to predict or reproduce with unit-tests.

- Hyper-simulator is not a replacement for your unit tests, you will still need them.
- Hyper-simulator might be a replacement for your friends, if you only keep them around as involuntary beta-testers.

Write a scenario for your application and let it run for a couple of minutes.
If you did **NOT** uncover any new errors within 5 minutes of sim-time,
please file a [formal complaint](/decentstack/hyper-simulator/issues/new).
<sub>(I'm happy to assit!)</sub>


Using the detailed `ndjson` output you can benchmark your program or
create an [interactive visualization](https://github.com/decentstack/hypersim-visualizer).
We also have built-in support for dumping simulator-output directly to Elasticsearch. Further offline analysis can be done anyway you desire


<details><summary><strong><a>How it works</a></strong> <sub>(Technical Details)</sub></summary>
It's pretty simple!

The whole simulation runs **offline** and _in-memory_.

It features a _simulated swarm_ and discovery process compatible with the `hyperswarm` API.

Each connection generated within the simulator is wrapped in a [ThrottledStream](https://github.com/decentstack/hyper-simulator/blob/master/lib/throttled-stream.js) that gives us control over data-transmission.

So the simulator runs in _iterations_ where each call to `sim.process(deltaTime)` generates a heartbeat/"tick"

It uses the same approach as many physics-based games, having a mainLoop that takes an amount of time as input which is used to fuel the calculations.

During each tick; we animate the simulated swarm, firing simulated `peer` and `connection` events,  and then pump all active ThrottledStreams providing them with bandwidth calculated by the amount of bandwidth remaining between the two peers. (the `QoS` could use some improvement.)

Each such process iteration generates multiple `ndjson` messages.
The `signal` method is a helper to let peers in your scenario report their own log messages during the run. :) (your message automatically get stamped with the peer-id nothing special)

The resulting logfile can be analyzed offline or in realtime but given some custom signals it can tell you how your application will behave given a network of N-Peers.

The project arose as an alternative to organize 30 people to run `your_application@version` and ask them to shout when the desired objective is reached. :)

Anyway, let me know if you give it a test-run, I've only used it to debug scaling issues in my own apps so far.

P.S.
The `storage` is convenient `random-access-file` wrapper that subpaths everything as `$PWD/.hypersim_cache/${peer.id}/${requested_path}`, the `_cache` folder gets deleted before and after each run.
</details>

<br/>

Features a built-in realtime log-aggregator:

[![asciicast](https://asciinema.org/a/9AcikNnlS2UA8yR8eY3MFRi2R.svg)](https://asciinema.org/a/9AcikNnlS2UA8yR8eY3MFRi2R)

Related projects:
- [hypersim-parser](https://github.com/decentstack/hypersim-parser) Fast and memory efficient ndjson aggregator. (Starting point for log-parsing)
- [hypersim-visualizer](https://github.com/decentstack/hypersim-visualizer) fancy Vue+D3.js visualizer.

## <a name="install"></a> Install

```bash
# with npm
npm i --save-dev hyper-simulator

# or with yarn
yarn add -D hyper-simulator

```

## <a name="usage"></a> Usage

Running from commandline:

```bash
# Run scenario and display the realtime stats in terminal
$(npm bin)/hypersim -T scenarios/myapp-scenario.js

# Run scenario and dump events to file
$(npm bin)/hypersim scenarios/myapp-scenario.js > swarm-log.json

# Run scenario and index all events in Elasticsearch
$(npm bin)/hypersim -e http://localhost:9200 scenario/myapp-scenario.js
```

### Defining a Scenario

Example scenario:
```js
const Simulator = require('hyper-simulator')

const sim = new Simulator()

await sim.ready() // Give the simulator some time to initialize

function MySeedSpawner() {
  // See peer-factory docs below.
}

function MyLeechSpawner() {
  // See peer-factory docs below.
}

// Launch 1 slow seed
sim.launch('seed',{ linkRate: 56<<8 }, MySeedSpawner())

// Launch 10 peers
for (let i = 0; i < 10; i++) {
  sim.launch('peer', MyLeechSpawner())
}

// Run simulation 1x speed and 200ms delay between ticks.
sim.run(1, 200)
  .then(() => console.error('Simulation finished'))
  .catch(err => console.error('Simulation failed', err)

// Simulator will run until manually stopped or all peers have signaled that they're done.
```

### Peer factory function

The launch interface is defined as followed:

#### `simulator.launch([name], [opts], initFn)`

- _optional_ **name** `String` tag the peer with a name or descriptor.
- _optional_ **opts** `Object` Launch options se below for defaults
- **intFn** `function(simContext, peerDone)` Peer factory function, see below
- **return** the result of `initFn` invocation.

Example opts:
```js
{
  name: 'anon',       // Tag peer with a name.
  linkRate: 102400,   // Maximum bandwidth in Bytes/Second
  maxConnections: 3,  // Maximum amount of active connections to maintain.
}
```

Example peer factory function:

```js
function SpawnPeer(context, peerDone) {
  // Destructure all values in context
  const {
    id,      // Uniquely generated id
    name,    // Name of peer as specified by launch()
    storage, // An instance of random-access-file,
    swarm,   // An instance of simulated hyperswarm
    opts,    // Copy of opts given to launch()
    signal,  // Function signal(eventName, payload) (more on this below)
    ontick,  // Lets you register a handler for peer.tick event: ontick(myHandlerFun)
    setTimeout, // Like vanilla setTimeout but uses simulated time instead.
    simulator // reference to the simulator.
  } = context

  // Initialize a decentralized feed or your application.
  const feed = hypercore(storage)

  // Find peers to replicate with
  feed.ready(() => {
    // Handle remote connections
    swarm.on('connection', (socket, { client }) => {
      socket
        .pipe(feed.replicate(client, { live: true }))
        .pipe(socket)
    })

    // Join the topic
    swarm.join(feed.key)
  })

  // Add custom signaling to the simulator output
  feed.on('download', seq => {
    signal('download', { seq })
  })

  // Attach peer.tick handler
  ontick(summary => {
    // Signal done when this peer has more than 5 blocks downloaded.
    if (feed.downloaded() > 5) peerDone()

    // Decorate the simulator peer#tick output with extra keys/values
    return { downloadedBlocks: feed.downloaded() }
  })
}

```

## Swarm-log

The following built-in events that are logged,

**Global keys**

All events will contain the following keys:

- **iteration** `{Number}` number of simulator ticks since start.
- **sessionId** `{Number}` Time/date at simulator initialization.
- **type** `{String}` Type of event
- **event** `{String}` Name of event
- **time** `{Number}` Amount of virtual milliseconds since simulator start.

#### simulator `state-initializing`

Emitted when simulator is instantiated

- **type** `simulator`
- **event** `state-initializing`
- **swarm** `hypersim|hyperswarm`


#### simulator `state-ready`

Emitted when simulator becomes ready

- **type** `simulator`
- **event** `state-ready`
- **swarm** `hypersim|hyperswarm`

#### simulator `state-running`
Emitted when simulator is started.
- **type** `simulator`
- **event** `state-running`
- **speed** `{number}` desc
- **interval** `{number}` desc

#### simulator `tick`

Emitted once on each tick

- **type** `simulator`
- **event** `tick`
- **state** `{string}` current state
- **delta** `{number}` virtual amount of milliseconds since previous tick.
- **pending** `{number}` Amount of non-finished peers in simulation.
- **connections** `{number}` Total amount of connections in swarm.
- **peers** `{number}` Amount of peers in swarm
- **capacity** `{number}` Sum of all peers' bandwidth
- **rate** `{number}` Sum of all peers' reported rate
- **load** `{number}` rate / capacity

#### simulator `state-finished`

Emitted once when all active peers have reached the 'done' or exited with an error.

- **type** `simulator`
- **event** `state-finished`


#### peer `init`

Emitted when a peer is launched

- **type** `peer`
- **event** `init`
- **id** `{number}` unique id of the peer.
- **name** `{string}` Peer name as provided during launch


#### peer `end`

Emitted when a peer has signaled being done or has thrown
an uncaught error.

- **type** `peer`
- **event** `end`
- **id** `{number}` unique id of the peer.
- **name** `{string}` Peer name as provided during launch
- **error** `{object}` null or the error that caused the peer to exit.

#### peer `tick`

Emitted by each peer every simulator tick.

- **type** `peer`
- **event** `tick`
- **id** `{number}` Unique peer id.
- **name** `{string}` Peer name
- **rx** `{number}` Number of bytes received during tick
- **tx** `{number}` Number of bytes transfered during tick
- **maxConnections** `{number}` Connection limit.
- **connectionCount** `{number}` Current connection count.
- **linkRate** `{number}` Bandwidth limit of peer.
- **load** `{number}` Bandwidth used during tick.
- **state** `{string}` State of peer. `active` or `done`
- **age** `{number}` Amount of ticks lived
- **finished** `{boolean}` true when peer signaled being done/ended.

#### socket `tick`

Emitted on each tick for each active connection in simulation.

- **type** `socket`
- **event** `tick`
- **rx** `{number}` Number of bytes received during tick
- **tx** `{number}` Number of bytes transfered during tick
- **noop** `{boolean}` true if no traffic occured during this tick
- **rxEnd** `{boolean}` receving end of socket is closed
- **txEnd** `{boolean}` transmitting end of socket is closed
- **rxDrained** `{boolean}` Receiving buffer not empty
- **txDrained** `{boolean}` Transmitting buffer not empty
- **id** `{string}` Unique id of socket
- **src** `{number}` Source peer id
- **dst** `{number}` Destination peer id
- **load** `{number}` (rx + tx) / Math.min(srcPeer.linkRate, dstPeer.linkRate)

#### custom

Emitted using the `signal` method provided by the peer launch interface.

- **type** `custom`
- **event** `{string}` The event name as signaled by application.
- **name** `{string}` Peer name that emitted the event
- **id** `{number}` Id of peer that emitted the event
- `YOUR CUSTOM METRICS`

## <a name="contribute"></a> Contributing

Ideas and contributions to the project are welcome. You must follow this [guideline](https://github.com/decentstack/hyper-simulator/blob/master/CONTRIBUTING.md).

## License

GNU AGPLv3 © Tony Ivanov



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.8.1] - 2020-01-08
### Changed
- Replaced Math.random() with deterministic alternative pulled from v8
- Fixed accidental logfile creation named `undefined`.
### Added
- TermAggregator: Transfers totals were added to peer display

## [0.9.0] - 2020-01-15
### Added
- Elasticsearch indexing support

## [1.0.0] - 2020-02-05
No more errors during prolonged runs!

### Added
- Deterministic pseudo `random()` is now available via peer context
- `context.timeout(Number) -> Promise` added. setTimeout() for Simulated time.
### Changed
- Fixed eternal loop bug caused by fifo.peek()
- Fixed issue with ThrottledStream where the readable state was closed twice.

## [1.1.0] - 2020-03-05
### Added
- Implemented BoundSwarm#leave(topic, cb)
### Changed
- Fixed invalid storage path. apologies for polluting your 'node_modules/'

## [1.2.0] - 2020-03-07
### Added
- SimulatedPeer#context.setTimeout() added
### Changed
- Replaced timer checker with new fast scheduler implmentation.
- Deprecated SimulatedPeer#context.timeout

## [1.2.1] - 2020-03-07
### Changed
- Fixed typo that caused premature invocation


test/index.js
const test = require('tape')
const hypercore = require('hypercore')
const HyperSim = require('..')
const { defer } = require('deferinfer')

function noop () {}

test('Discovery & Transmission', async t => {
  t.plan(6)
  try {
    const sim = new HyperSim({
      logger: noop // line => console.error(JSON.stringify(line))
    })
    await sim.ready()

    sim.launch(({ swarm, signal, name }, end) => {
      let pending = 1

      swarm.join(Buffer.from('stub-topic'), {
        lookup: false, // find & connect to peers
        announce: true // optional- announce self as a connection target
      })

      swarm.once('connection', (socket, details) => {
        t.pass('seed onConnection: ' + socket.id)
        socket.once('close', () => {
          // t.equal(detail.client, false, 'Initiating boolean available')
          if (!--pending) t.notOk(end(), 'Seed stream closed')
        })

        socket.on('data', chunk => {
          t.equal(chunk.toString(), 'hey seed', 'Leech msg received')
          socket.write('Yo leech!')
          socket.end()
        })
      })
    })

    sim.launch(({ swarm, signal, name }, end) => {
      swarm.join(Buffer.from('stub-topic'), {
        lookup: true, // find & connect to peers
        announce: true // optional- announce self as a connection target
      })
      swarm.once('connection', (socket, details) => {
        t.pass('leech onConnection ' + socket.id)
        socket.once('data', chunk => {
          t.equal(chunk.toString(), 'Yo leech!', 'Seed msg received')
          socket.end()
          // socket.destroy()
        })
        socket.once('close', () => {
          t.notOk(end(), 'Leech stream closed')
        })
        socket.write('hey seed')
      })
    })

    await sim.run(2, 100)
    t.end()
  } catch (err) { t.error(err) }
})

test('Basic hypercore simulation', async t => {
  t.plan(5)
  try {
    const sim = new HyperSim({
      logger: noop
      // logger: HyperSim.TermMachine()
      // logger: line => console.error(line)
    })

    await sim.ready()

    const initSeedFn = ({ storage, swarm }, end) => {
      return defer(done => {
        const feed = hypercore(storage)
        feed.ready(() => {
          feed.append('hello world!', err => {
            t.error(err)
            swarm.join(feed.key, { lookup: false })
            swarm.on('connection', (socket, { client }) => {
              socket
                .pipe(feed.replicate(client, { live: true }))
                .pipe(socket)
                .on('error', t.error)
            })
            end()
            done(null, feed)
          })
        })
      })
    }

    // Launch 1 nerfed seed
    const seedFeed = await sim.launch('seed',
      { linkSpeed: 50 * 1024, maxConnections: 1 },
      initSeedFn
    )

    // Launch 3 leeches
    for (let i = 0; i < 2; i++) {
      sim.launch('leech', ({ storage, swarm, signal }, end) => {
        const feed = hypercore(storage, seedFeed.key)
        feed.ready(() => {
          swarm.join(feed.key)
          swarm.on('connection', (socket, { client }) => {
            socket
              .pipe(feed.replicate(client, { live: true }))
              .pipe(socket)
              .on('error', t.error)
          })
          feed.get(0, (err, chunk) => {
            t.error(err)
            signal('msg-replicated')
            t.equal(chunk.toString('utf8'), 'hello world!', `Leech${i} done`)
            end()
          })
        })
      })
    }

    // Run the simulation
    await sim.run(2, 100)
  } catch (err) { t.error(err) }
  t.end()
})

/*
 * Debugging streamx-states, turns out that a writable.push(null)
 * manifests itself as the _final() callback. This of course
 * needed to be translated into a readable.push(null) when ducttaping
 * two duplex-streams together. Apologies for littering but i'm leaving
 * this here for future personal refernce:
 *
> protoStream._duplexState.toString(2).padStart(32,'0')
4333824
'00000000010000100010000100000000'
                   READ_DONE
> socket._duplexState.toString(2).padStart(32,'0')
13336848
'00000000110010111000000100010000'
                            Primary Read

> protoStream._duplexState
  .toString(2).split('').reverse().map((i,n) => parseInt(i) && n).filter(n => n)
[ 8,  // READ_PIPE_DRAINED
  13, // READ_DONE
  17, // WRITE_PRIMARY
  22  // WRITE_EMIT_DRAIN
]

> socket._duplexState.toString(2).split('').reverse().map((i,n) => parseInt(i) && n).filter(n => n)
[ 4, // READ_PRIMARY
  8, // READ_PIPE_DRAINED
  15,// READ_NEEDS_PUSH
  16,// WRITE_ACTIVE
  17,// WRITE_PRIMARY
  19,// WRITE_QUEUED
  22,// WRITE_EMIT_DRAIN
  23 // WRITE_NEXT_TICK (also assumes WRITE_ACTIVE)]

*/


test/scheduler.js
const test = require('tape')

const Scheduler = require('../lib/scheduler')

test('enque && tick', t => {
  t.plan(5)
  const s = new Scheduler()
  let n = 0
  s.enque(5, () => t.equals(++n, 2, 'Second'))
  s.enque(3, () => t.equals(++n, 1, 'First'))
  s.enque(15, () => { t.equals(++n, 4, 'Fourth'); t.end() })
  s.enque(7, () => t.equals(++n, 3, 'Third'))

  t.equals(s.toArray().map(i => i.at).join(','), '3,5,7,15')
  let i = 0
  while (!s.empty) s.tick(i++)
})

test('cancel', t => {
  t.plan(4)
  const s = new Scheduler(false)
  let n = 0
  s.enque(2, () => t.equals(++n, 1, 'First'))
  const cancel = s.enque(4, t.fail.bind(null, 'Should have been canceled'))
  s.enque(13, () => t.equals(++n, 3, 'Third'))
  s.enque(5, () => t.equals(++n, 2, 'Second'))
  cancel()

  t.equals(s.toArray().map(i => i.at).join(','), '2,5,13')
  let i = 0
  while (!s.empty) s.tick(i++)
})


index.js
// SPDX-License-Identifier: AGPL-3.0-or-later
const raf = require('random-access-file')
const { defer, infer } = require('deferinfer')
const { mkdirSync, statSync, createWriteStream } = require('fs')
const { join } = require('path')
const rimraf = require('rimraf')
const { EventEmitter } = require('events')
const hyperswarm = require('hyperswarm')
const BufferedThrottleStream = require('./lib/throttled-stream')
const termAggregator = require('./lib/term-aggregator')
const ElasticStreamer = require('./lib/elastic-streamer.js')
const prand = require('./lib/pseudo-random.js')
const Scheduler = require('./lib/scheduler')

// const sos = require('save-our-sanity')

const INIT = 'initializing'
const READY = 'ready'
const RUNNING = 'running'
const FINISHED = 'finished'

// Global socket counter
let sckCtr = 0

class SimulatedPeer {
  constructor (id, signal, opts = {}) {
    this.id = id
    this.name = opts.name
    this.linkRate = opts.linkRate || 102400
    // this.latency = opts.latency || 100 // Not used yet
    this.maxConnections = opts.maxConnections || 3
    this.sockets = []
    this.lastTick = 0
    this.age = 0
    this._consumed = { rx: 0, tx: 0 }
    this.finished = false
    this._signal = (ev, payload = {}) => signal('peer', ev, {
      id,
      name: this.name,
      ...payload
    })
    this._rootSignal = signal
    this._signal('init')
    this.scheduler = new Scheduler()
    this.version = null
  }

  isConnected (peer) {
    for (const { src, dst } of this.sockets) {
      if (src.id === peer.id || dst.id === peer.id) return true
    }
    return false
  }

  connect (peer, callback) {
    if (this.isPoolFull || peer.isPoolFull) return callback(new Error('Connection Pool Exhausted'))
    // Maybe just return false and silently abort on already connected.
    if (this.isConnected(peer)) return callback(new Error('Already connected'))
    const socket = new BufferedThrottleStream()
    socket.id = `${sckCtr++}` // #${this.id}:${peer.id}`
    socket.src = this
    socket.dst = peer
    const x = this._accept(socket, true)
    const y = peer._accept(socket, false)
    if (!(x && y)) throw new Error('Invalid State') // This shouldn't ever happen
    return callback(null, socket)
  }

  get hasSimSwarm () {
    return this.boundSwarm && this.boundSwarm.__HYPERSIM__
  }

  tick (iteration, delta) {
    this.age++
    this.lastTick = iteration

    const { rx, tx } = this._consumed
    this._consumed = { rx: 0, tx: 0 }

    // tick BoundSwarm
    if (this.hasSimSwarm) this.boundSwarm.tick(iteration, delta)
    const summary = {
      rx,
      tx,
      maxConnections: this.maxConnections,
      connectionCount: this.sockets.length,
      linkRate: this.linkRate,
      load: (rx + tx) / (this.linkRate * delta / 1000),
      state: this.finished ? 'done' : 'active',
      age: this.age,
      finished: this.finished,
      version: this.version
    }

    // Run scheduler
    this.scheduler.deltaTick(delta)

    // Call handler and merge it's output into summary
    if (typeof this._onTickHandler === 'function') {
      const output = this._onTickHandler(summary)
      if (typeof output === 'object' && output !== null) Object.assign(summary, output)
    }
    // log Peer summary
    this._signal('tick', summary)
  }

  get isPoolFull () {
    return !(this.sockets.length < this.maxConnections)
  }

  _inccon (rx, tx) {
    this._consumed.rx += rx
    this._consumed.tx += tx
  }

  _end (error) {
    if (this.finished) return console.error('WARNING!: `end` invoked more than once by:', this.name, this.id)
    this.finished = true
    // TODO: unregister peer from simulator.
    // this.peers.splice(this.peers.indexOf(peer), 1)
    this._signal('end', { error })
  }

  _removeSocket (socket) {
    const idx = this.sockets.indexOf(socket)
    if (idx !== -1) this.sockets.splice(idx, 1)
    // this._signal('disconnect', {
    //   src: this.peer.id,
    //   dst: this.remotePeer.id,
    //   sid
    // })
  }

  _accept (socket, initiator) {
    if (this.isPoolFull) return false
    this.sockets.push(socket)
    return true
  }

  _bootstrap (simulator, storage, simdht, initFn, opts) {
    this.storage = storage

    if (simdht) {
      this.simdht = simdht
      this.boundSwarm = new BoundSwarm(simdht, this)
    } else {
      // TODO: alternatively support a real instance of hyperswarm
      this.boundSwarm = hyperswarm()
    }

    // Construct the Peer Context,
    // the primary interface to the simulator.
    const context = {
      opts,
      id: this.id,
      name: this.name,
      storage,
      swarm: this.boundSwarm,
      simulator,
      signal: (ev, args = {}) => {
        this._rootSignal('custom', ev, { ...args, name: this.name, id: this.id })
      },
      ontick: handler => { this._onTickHandler = handler },
      random: prand,
      setTimeout: this.scheduler.setTimeout.bind(this.scheduler),
      timeout: msec => {
        console.warning('deprecated: Peer#context.timeout, use Peer#context.setTimeout')
        return new Promise(resolve => this.scheduler.setTimeout(msec, resolve))
      },
      trace: () => traceNetwork(this)
    }
    Object.defineProperty(context, 'version', {
      get: () => this.version,
      set: v => { this.version = v }
    })
    let peerDone = null
    const prom = defer(done => { peerDone = done })

    const pass = initFn(context, err => {
      this._end(err || null)
      // this._swarm._destoyConnectionsOf(id)
      peerDone()
    })

    return { pass, prom }
  }
}

class BoundSwarm extends EventEmitter {
  get __HYPERSIM__ () { return true }
  constructor (dht, peer) {
    super()
    this.dht = dht
    this.peer = peer
    this.topics = new Map()
    // Ambitious, but i'd also like to model dropping
    // connections mid-replicate and observe how that affects the swarm
    // finding out decentralized services perform in bad link conditions.
    this.connectFailureRate = 0
    this.lastLookup = 0
    this.peerQueue = []
    this.banned = []
    this.joinedTopics = []
    this.age = 0
  }

  topic (key) {
    this.topics.get(key)
  }

  tick (iteration, delta) {
    this.age += delta
    // Fetch new peers
    if (this.lookup && this.age - this.lastLookup > 1000) {
      this.queryLookup()
    }

    // Attempt to connect
    if (this.lookup &&
      this.peerQueue.length &&
      this.peer.sockets.length < this.peer.maxConnections) {
      const que = this.peerQueue
      this.peerQueue = []
      que.forEach(n => {
        if (this.age - n.lastAttempt > 5000) {
          n.lastAttempt = this.age
          const { peer } = n
          // Attempt to connect
          this.peer.connect(peer, (err, socket) => {
            if (err) return this.peerQueue.push(n)
            const topic = n.topic[0] === '¤' ? Buffer.from(n.topic.substr(1), 'hex') : n.topic
            socket.topic = topic
            this.emit('connection', socket, { client: true, topic })
            peer.boundSwarm.emit('connection', socket.out, { client: false, topic })

            // this.peer.signal(initiator ? 'connect' : 'accept', {
            //   src: this.peer.id,
            //   dst: peer.id,
            //   sid: socket.id
            // })
          })
        } else if (--n.attempts) {
          this.peerQueue.push(n)
        } else this.banned.push(n.peer.id)
      })
    }
  }

  join (topic, { lookup, announce } = {}) {
    if (Buffer.isBuffer(topic)) topic = '¤' + topic.toString('hex') // buffers can't be keys
    this.lookup = !(lookup === false)
    this.announce = !(announce === false)

    if (!this.lookup && !this.announce) throw new Error('At both lookup and announce can\'t be set to false')
    if (this.lookup) this.queryLookup()
    if (this.announce) this.doAnnounce(topic)
  }

  leave (topic, cb) {
    if (Buffer.isBuffer(topic)) topic = '¤' + topic.toString('hex') // buffers can't be keys
    this.dht.unregister(topic, this.peer)
    this.joinedTopics.splice(this.joinedTopics.indexOf(topic), 1)
    if (typeof cb === 'function') cb()
  }

  doAnnounce (topic) {
    this.joinedTopics.push(topic)
    this.dht.register(topic, this.peer)
  }

  queryLookup () {
    for (const topic of this.joinedTopics) {
      for (const cand of this.dht.getCandiates(topic)) {
        if (this.banned.indexOf(cand.id) === -1 &&
          cand.id !== this.peer.id) {
          this.peerQueue.push({
            lastAttempt: -20000,
            attempts: 10,
            peer: cand,
            topic
          })
          this.emit('peer', cand)
        }
      }
    }
  }
}

class SimDHT {
  constructor (signal) {
    this.listeningPeers = {}
    this.signal = signal
  }

  register (topic, record) {
    topic = topic.toString()
    this.listeningPeers[topic] = this.listeningPeers[topic] || []
    this.listeningPeers[topic].push(record)
  }

  unregister (topic, record) {
    const records = this.listeningPeers[topic]
    const idx = records.indexOf(record)
    if (idx !== -1) records.splice(idx, 1)
  }

  getCandiates (topic) {
    return this.listeningPeers[topic]
      .sort(() => prand() - 0.5) // shuffle
  }

  tick (iteration, deltaTime) {
    // TODO: auto remove stale
  }

  async close () {
    // TOOD: Not Implemented / Not needed?
  }
}

class Simulator extends EventEmitter {
  constructor (opts = {}) {
    super()
    Object.assign(opts, env2opts())

    this.poolPath = opts.poolPath || join(process.cwd(), '.hypersim_cache')
    this._idCtr = 0
    this._simdht = new SimDHT(this._signal.bind(this))

    // Default logging behaviour is to output ndjson to STDOUT
    this._logger = opts.logger || (msg => console.log(JSON.stringify(msg)))
    if (opts._OnFinishHack) this.on(FINISHED, opts._OnFinishHack)

    this.time = 0
    this.iteration = 0
    this.sessionId = new Date().getTime()
    this.pendingPeers = 0
    this.finishedPeers = 0
    this.run = this.run.bind(this)
    this.peers = []
    this.bwReserve = 10 << 8 // Minimum bandwidth for socket per tick 10KiloByte
    this._transition(INIT, {
      swarm: 'hypersim'
    })
  }

  _transition (newState, ...extra) {
    this._signal('simulator', `state-${newState}`, ...extra)
    this.emit(newState)
    this._state = newState
  }

  ready (cb) {
    if (!this._readyResult) {
      this._readyResult = defer(async done => {
        // Purge any left overs
        await this._purgeStorage()

        // Initialize new fold structure
        try {
          const stat = statSync(this.poolPath)
          if (!stat.isDirectory()) throw new Error(`Path already exists and is not a directory: ${this.poolPath}`)
          this._signal('simulator', 'using-cache', { path: this.poolPath })
        } catch (err) {
          if (err.code !== 'ENOENT') throw err
          mkdirSync(this.poolPath)
          this._signal('simulator', 'create-cache', { path: this.poolPath })
        }
        this._transition(READY, {})
        done(null)
      })
    }
    return infer(this._readyResult, cb)
  }

  launch (name = 'anon', opts = {}, initFn) {
    if (typeof name !== 'string') return this.launch(undefined, name, opts)
    if (typeof opts === 'function') return this.launch(name, undefined, opts)
    if (typeof initFn !== 'function') throw new Error('Missing parameter initFn, must be a function')
    opts.name = name
    const id = ++this._idCtr
    const storage = f => raf(join(this.poolPath, String(id), f))
    // TODO: implement random-access-metered
    const peer = new SimulatedPeer(id, this._signal.bind(this), opts)

    this.pendingPeers++
    this.peers.push(peer)
    const { pass, prom } = peer._bootstrap(this, storage, this._simdht, initFn, opts)
    prom.then(() => this.finishedPeers++)

    return pass
  }

  _signal (type, ev, payload = {}) {
    this._logger({
      type,
      event: ev,
      ...payload,
      time: this.time,
      iteration: this.iteration,
      sessionId: this.sessionId
    })
  }

  run (speed = 0.1, interval = 200) {
    this._transition(RUNNING, { speed, interval })
    return defer(async done => {
      // Why am i measuring real time between ticks?
      // they should be running in sim time..
      let timeLastRun = new Date().getTime()
      while (this._state === RUNNING) {
        await defer(d => setTimeout(d, interval))
        const time = new Date().getTime()
        const deltaTime = (time - timeLastRun) * speed
        this._tick(deltaTime)
        timeLastRun = time
      }
      done()
    }).then(() => this.teardown())
  }

  _tick (deltaTime) {
    if (this.pendingPeers && this.pendingPeers - this.finishedPeers < 1) this._transition(FINISHED)
    const iteration = ++this.iteration
    this.time += deltaTime

    this._simdht.tick(iteration, deltaTime) // if is hyperswarmSim

    const summary = {
      state: this._state,
      delta: deltaTime,
      pending: this.pendingPeers - this.finishedPeers,
      connections: 0,
      peers: this.peers.length
    }

    let consumedBandwidth = 0
    const budgets = {}

    const swarmBandwidthCapacity = this.peers.reduce((sum, p) => {
      const h = p.sockets.length * this.bwReserve
      budgets[p.id] = (p.linkRate - h) * deltaTime / 1000
      return sum + budgets[p.id]
    }, 0)

    // Pump the simulated sockets
    for (const peer of this.peers) {
      // Yep, we have a QoS issue, reserving bandwith is one way but
      // but each tick we exhaust the bandwith on the first socket..
      // TODO: Improve this quick'n'dirty random shuffle.
      for (const socket of peer.sockets.sort(() => prand() - 0.5)) {
        const { src, dst } = socket
        if (socket.lastTick >= iteration) continue
        const budget = Math.max(this.bwReserve * deltaTime / 1000, Math.min(budgets[src.id], budgets[dst.id]))
        // We're simplifying budgets for sockets, treating them
        // as if they're all half-duplex.
        const stat = socket.tick(iteration, budget)
        const { rx, tx, noop, rxEnd, txEnd, destroyed } = stat
        if (rxEnd || destroyed) src._removeSocket(socket)
        if (txEnd || destroyed) dst._removeSocket(socket)
        if (!noop) summary.connections++
        const load = (rx + tx) / budget

        this._signal('socket', 'tick', {
          ...stat,
          id: socket.id,
          src: src.id,
          dst: dst.id,
          load
        })

        budgets[src.id] -= rx + tx
        budgets[dst.id] -= rx + tx
        consumedBandwidth += rx + tx
        src._inccon(rx, tx)
        dst._inccon(tx, rx)
      }
      peer.tick(iteration, deltaTime)
    }
    summary.capacity = swarmBandwidthCapacity
    summary.rate = deltaTime ? Math.ceil(consumedBandwidth / (deltaTime / 1000)) : -1
    if (swarmBandwidthCapacity > 0) summary.load = summary.rate / swarmBandwidthCapacity
    // TODO: add memory consumption to metrics
    this.emit('tick', iteration, summary)
    this._signal('simulator', 'tick', summary)
  }

  _purgeStorage () {
    return defer(done => rimraf(this.poolPath, done))
  }

  async teardown () {
    await this._simdht.close()
    await this._purgeStorage()
    /*
    for (const peer of this.peers) {
      await peer.close()
    } */
  }
}

function env2opts () {
  const opts = {}
  let logger = null

  if (process.env.HYPERSIM_OUT) {
    const outputFile = process.env.HYPERSIM_OUT
    const logFileStream = createWriteStream(outputFile)
    logger = line => logFileStream.write(JSON.stringify(line))
  } else if (process.env.HYPERSIM_ELASTIC_URL) {
    const estream = new ElasticStreamer(process.env.HYPERSIM_ELASTIC_URL)
    logger = line => estream.index(line)
    opts._OnFinishHack = () => {
      estream.commitBatch(true)
    }
  }

  if (process.env.HYPERSIM_TEE) {
    const l1 = logger || (() => {})
    const l2 = termAggregator()
    logger = line => {
      l1(line)
      l2(line)
    }
  }

  if (logger) opts.logger = logger
  return opts
}

// Slow recursive function for diagnostics.
// I wanted to make sure that the random network topology
// is not what causing forward failure
function traceNetwork (peer, path = []) {
  if (!path.length) path.push(peer) // store first

  for (const socket of peer.sockets) {
    const other = socket.dst === peer ? socket.src : socket.dst
    if (path.indexOf(other) === -1) {
      path.push(other) // new peer discovered
      return traceNetwork(other, path)
    }
  }
  if (path[0] === peer) return path
  const prev = path[path.indexOf(peer) - 1]
  return traceNetwork(prev, path)
}

module.exports = Simulator
module.exports.BufferedThrottleStream = BufferedThrottleStream
module.exports.TermMachine = termAggregator
